pkg/dab/find.go

```diff
@@ -0,0 +1,237 @@
+// and if the 'searchUp' parameter is true, then each segment of 'searchPath'
+// will also be popped and searched.
+// (If 'searchUp' is false, the distinction between 'basisPath' and 'searchPath' vanishes.)
+//
+// If `{basisPath}/{searchPath}` contains a file, only file is inspected,
+// and search up through parent directories is not performed regardless of the value of 'searchUp'.
+//
+// The typical values of 'basisPath' and 'searchPath' vary depending on feature, but are usually one of:
+// basisPath is the CWD, and searchPath was a CLI argument;
+// or, basisPath is empty string (meaning the root of the filesystem), and searchPath is the CWD (possibly joined with a CLI argument);
+// or, basisPath is a workspace directory, and searchPath is the CWD within that (possibly joined with a CLI argument).
+// Neither value should ever have a leading slash (as is typical for APIs using FS).
+//
+// The 'accept' parameter can be used to have the function ignore (or return errors) in some scenarios.
+// For example, 'ActionableSearch_Any' may cause the search to return a formula or a module;
+// whereas if you only want only want to detect modules, you should use 'ActionableSearch_Module' instead.
+// See the comments on the constants for details on other options.
+//
+// It is possible for a module and plot to be returned;
+// or a module alone; or a plot alone; or a formula alone;
+// or an error; or all four of them may be nil at once if the search found nothing.
+// The foundPath and remainingSearchPath values are always returned,
+// even in the case of errors.
+//
+// Errors:
+//
+//  - warpforge-error-datatoonew -- when found data is not supported by this version of warpforge
+//  - warpforge-error-invalid-argument -- when a file contains the wrong content for its filename
+//  - warpforge-error-io -- when an IO error occurs during search
+//  - warpforge-error-missing -- when an expected file is missing
+//  - warpforge-error-module-invalid -- when a read module contains invalid data
+//  - warpforge-error-searching-filesystem -- when the search of the filesystem produces an invalid result
+//  - warpforge-error-serialization -- when IPLD deserialization fails
+func FindActionableFromFS(
+	fsys fs.FS,
+	basisPath string, searchPath string, searchUp bool,
+	accept ActionableSearch,
+) (
+	m *wfapi.Module, p *wfapi.Plot, f *wfapi.Formula,
+	foundPath string, remainingSearchPath string, err error,
+) {
+	remainingSearchPath = searchPath
+
+	// First round: this may be a file, and may be one of any of the types.
+	// Stat it and consider that first:
+	// if it is a file, that's its own whole detection procedure (and means no further search);
+	// if it's not a file, we can procede to the behavior for dir feature detection (without popping).
+	foundPath = filepath.Join(basisPath, remainingSearchPath)
+	fi, e2 := stat(fsys, foundPath)
+	if e2 != nil {
+		err = e2
+		return
+	}
+	if fi.Mode()&^fs.ModePerm == 0 {
```

Comment:
This logic feels ambiguous. `&^` is bitwise-AND _and_ bitwise-XOR? I don't know what this means.

Reply:
I have no idea what this is doing, @warpfork wrote it... can you explain?

Reply:
I looked it up and it's a bit clear (AND NOT) operation. Feels somewhat special and contrarian to lack a bitwise NOT operator. It's possible to do a bitwise NOT but it's an XOR NEGATE `(^-x)` so a bit clear would be `a ^ (^-b)` which... yeah bit clear needs to not be that. No idea why they didn't use `&~` but whatever.
```
package main

import "fmt"

func main() {
	a := 0b1010
	b := 0b0011
	fmt.Printf("       a = %#04b\n", a)
	fmt.Printf("       b = %#04b\n", b)
	fmt.Printf("   a ^ b = %#04b\n", a^b)
	fmt.Printf("   a & b = %#04b\n", a&b)
	fmt.Printf("  a &^ b = %#04b\n", a&^b)
	fmt.Printf("a ^(^-b) = %#04b\n", a^(^-b))
}
```
```
       a = 0b1010
       b = 0b0011
   a ^ b = 0b1001
   a & b = 0b0010
  a &^ b = 0b1000
a ^(^-b) = 0b1000

Program exited.
```
https://go.dev/play/p/3r_EypT_FHD

Reply:
Yeah this syntax surprised me to rediscover when I wrote it too, heh.

Overall what this means is "is it a normal file".  I wish there was a clearer way to write this (and if you can think of one, please go for it).

Reply:
Why not use `fi.Mode().IsRegular()`?

Reply:
... yep, that'd be better.  Dunno how I missed it.

Reply:
Tests still open irregular files during search mode. I don't really want to think about that right now so :shrug:.